### ISSUE TITLE
Hang on exit when not consuming `run_udf_iter` generator

### ISSUE THREAD

**Issue by sk1p** [LiberTEM/LiberTEM#1444]
The following test code causes the Python interpreter to not properly exit, and hangs instead, waiting for a background thread to exit:

```python
def test_hang(lt_ctx, default_raw):
    it = lt_ctx.run_udf_iter(dataset=default_raw, udf=[SumSigUDF()])
    for i in it:
        raise Exception()
```

Checking via `py-spy dump -p $(pgrep -f pytest)`:

```
Process 3665652: /home/alex/miniconda/envs/libertem/bin/python /home/alex/miniconda/envs/libertem/bin/pytest -s tests/test_context.py::test_hang
Python v3.9.7 (/home/alex/miniconda/envs/libertem/bin/python3.9)

Thread 3665652 (idle): "MainThread"
    _shutdown (threading.py:1448)
Thread 3665864 (idle): "Thread-1"
    wait (threading.py:312)
    get (queue.py:171)
    get (libertem/common/executor.py:471)
    __enter__ (contextlib.py:119)
    monitor_queue (libertem/common/progress.py:63)
    run (threading.py:910)
    _bootstrap_inner (threading.py:973)
    _bootstrap (threading.py:930)
```

So, this is caused by the progress monitor thread not being cleaned up properly. The thread bringup/cleanup code lives here:

https://github.com/LiberTEM/LiberTEM/blob/f2a71e82a6b81dffe6ad98f5e7ddb142bf861f55/src/libertem/common/progress.py#L33-L51

And is used, for example, here:

https://github.com/LiberTEM/LiberTEM/blob/f2a71e82a6b81dffe6ad98f5e7ddb142bf861f55/src/libertem/executor/inline.py#L72-L80

So, in theory, using the context manager should stop the thread and allow a clean exit. But, if the generator doesn't run to completion, for example because of an exception in _consumers_ of the generator, the `__exit__` is not called (until the generator is garbage collected).

There are a few "workaround" for this issue:

1) Run the generator to it's conclusion:

```python
def test_no_hang_1(lt_ctx, default_raw):
    it = lt_ctx.run_udf_iter(dataset=default_raw, udf=[SumSigUDF()])
    try:
        for i in it:
            raise Exception()
    finally:
        for i in it:
            pass
```

2) Use `contextlib.closing` on the generator:

```python
def test_no_hang_2(lt_ctx, default_raw):
    it = lt_ctx.run_udf_iter(dataset=default_raw, udf=[SumSigUDF()])
    with contextlib.closing(it) as it:
        for i in it:
            raise Exception()
```

3) ... which is roughly the same as explicitly calling `close` on the generator:

```python
def test_no_hang_3(lt_ctx, default_raw):
    it = lt_ctx.run_udf_iter(dataset=default_raw, udf=[SumSigUDF()])
    try:
        for i in it:
            raise Exception()
    finally:
        it.close()
```

4) ... and to show that `__exit__` is eventually called, on gargabe collection:

```python
def test_no_hang_4(lt_ctx, default_raw):
    it = lt_ctx.run_udf_iter(dataset=default_raw, udf=[SumSigUDF()])
    try:
        for i in it:
            raise Exception()
    finally:
        del it
        import gc; gc.collect()
```

(I think what's happening here is that the threading atexit code waits for the thread to exit, and this happens before the final garbage collection run, but the `__exit__` would only be called during said gc run...)

Quoting from https://peps.python.org/pep-0533/:

> But there’s a subtlety here, caused by the interaction of with blocks and generators. with blocks are Python’s main tool for managing cleanup, and they’re a powerful one, because they pin the lifetime of a resource to the lifetime of a stack frame. But this assumes that someone will take care of cleaning up the stack frame… and for generators, this requires that someone close them.

> In this case, adding the with block is enough to shut up the ResourceWarning, but this is misleading – the file object cleanup here is still dependent on the garbage collector. The with block will only be unwound when the read_newline_separated_json generator is closed. If the outer for loop runs to completion then the cleanup will happen immediately; but if this loop is terminated early by a break or an exception, then the with block won’t fire until the generator object is garbage collected.

Sadly, said PEP was not implemented and we don't have `__iterclose__` or similar available. How should we generally handle this case? Requiring users to `close` the generator or use the `closing` context manager would be the most robust solution for now; for this concrete issue, we could also have the monitoring thread be a daemon thread, which is cleaned up automatically on interpreter exit:

```patch
diff --git a/src/libertem/common/progress.py b/src/libertem/common/progress.py
index 360d7e15..9389f539 100644
--- a/src/libertem/common/progress.py
+++ b/src/libertem/common/progress.py
@@ -34,9 +34,12 @@ class CommsDispatcher:
         if self._thread is not None:
             raise RuntimeError('Cannot re-enter CommsDispatcher')
         self._thread = threading.Thread(target=self.monitor_queue)
+        self._thread.daemon = True
         self._thread.start()
```

**Comment by matbryan52** [comment]
Good catch, well traced! My bad for not considering this during the design of the progress monitor.

> we could also have the monitoring thread be a daemon thread, which is cleaned up automatically on interpreter exit

This makes sense to me in any outcome (based on my understanding of daemon threads) as the monitor is not performing any function except routing messages. If we lose some progress bar iterations during error situations I think that's alright?

My question is if making the monitor thread a daemon is sufficient? I just did some quick tests and it seems to allow us to catch the error during `run_udf_iter` and then do more work after, and the monitor thread is indeed cleaned up correctly.

It would certainly be a less heavy solution than requiring `closing()` blocks in user code!

**Comment by sk1p** [comment]
> Good catch, well traced! My bad for not considering this during the design of the progress monitor.

No worries - I find this aspect of generators a bit surprising, and also didn't ask myself the question "what happens if the generator is not consumed" during review...

> This makes sense to me in any outcome (based on my understanding of daemon threads) as the monitor is not performing any function except routing messages. If we lose some progress bar iterations during error situations I think that's alright?

Yeah, should be alright - I think it would be mostly critical if we had any persistent state (temporary files, lock files, stuff like that) that we need to clean up from that thread, which we don't.

> My question is if making the monitor thread a daemon is sufficient?

For trying this, I made the thread a daemon thread in #1441, and at least locally I get another hang-issue related to this, which I'm still debugging. I'll update this issue when I find the cause, and probably extract into a separate PR.

> I just did some quick tests and:
> 
> ```python
>     it = lt_ctx.run_udf_iter(dataset=default_raw, udf=[SumSigUDF()])
>     try:
>         for i in it:
>             raise Exception()
>     except:
>         ...
> ```
> 
> does not hang for me, even if I don't explicitly drain the iterator. The only place I can make it hang is if I have no error handling at all - i.e. the interpreter is going to shut down for the unhandled exception - and in this case using a daemon thread solves the problem by not blocking.

I hit this issue when running `pytest` - when doing TDD, the first iterations fail by design (example: the method you call in the test is not implemented yet, and running the test confirms that you are hitting the correct code paths). It's possible that the way exceptions are caught in this case ultimately triggers the issue?

**Comment by sk1p** [comment]
> at least locally I get another hang-issue related to this, which I'm still debugging. I'll update this issue when I find the cause, and probably extract into a separate PR.

This issue was unrelated, and caused by not properly forwarding a `generator.throw(...)` to the right upstream generators. This caused basically the same issue - the inner generator was not running to completion, and also wasn't interrupted by an exception. As a consequence, a second `CommsDispatcher` thread for the next `run_udf` call was started before the old one was finished, causing the `'STOP'` message to end up on the wrong thread.

This was fixable by properly forwarding the exception to the innermost generator - in general, the `PipelinedExecutor` would need some changes to support multiple running jobs concurrently, which would need to multiplex communication for multiple `cancel_id`s over the same `message_q`, like we do with the dask executor.
